### PR TITLE
fix(runtime): add smart GPU allocation to prevent multi-model OOM crashes

### PR DESCRIPTION
**Pull request by mhamann**
## Summary

- Adds a GPU allocator that queries actual free VRAM per device and routes each model to the single GPU with the most headroom (split_mode=NONE), falling back to multi-GPU split only when necessary
- Fixes context_calculator to use `torch.cuda.mem_get_info()` (free memory) instead of `get_device_properties().total_memory`, and supports per-GPU queries via new `gpu_index` parameter
- Exposes `split_mode` parameter in llamafarm_llama's `Llama` class so callers can control GPU splitting behavior
- Adds `InsufficientVRAMError` that propagates as a clean RuntimeError instead of crashing the process

## Changes

- **New: `gpu_allocator.py`** — GPU enumeration, VRAM estimation (weights + KV cache + overhead), and allocation logic with three strategies: single-GPU, multi-GPU proportional split, or error
- **`llama.py`** — Exposed `split_mode` parameter (0=NONE, 1=LAYER, 2=ROW)
- **`context_calculator.py`** — `get_available_memory()` now returns free VRAM on specified GPU; `get_default_context_size()` accepts `gpu_index`
- **`gguf_language_model.py`** — Integrates GPU allocator before model load, re-computes context size using allocated GPU's actual free memory
- **`gguf_encoder_model.py`** — Same GPU allocator integration for embedding models
- **`device.py`** — `get_device_info()` reports per-GPU free/total memory and gpu_count for multi-GPU systems
- **New: `test_gpu_allocator.py`** — 19 tests including the exact RTX 3090 + Quadro M4000 scenario
- **`test_context_calculator.py`** — Updated to test `mem_get_info` and `gpu_index` parameter

## Test Plan

- [x] All 44 tests in `test_gpu_allocator.py` and `test_context_calculator.py` pass
- [x] All 23 tests in `test_gguf_model.py` and `test_gguf_encoder_model.py` pass (no regressions)
- [ ] Manual test on multi-GPU system: load two GGUF models, verify second model loads on GPU with most free VRAM

---

> [!NOTE]
> **Medium Risk**
> Touches GPU placement and context sizing logic during model load, which can change runtime behavior across multi-GPU setups and may surface edge-case OOM/perf regressions despite added tests.
> 
> **Overview**
> Adds a new **VRAM-aware GPU allocator** for GGUF inference that queries per-device *free* memory, estimates per-model VRAM needs (weights + KV cache + overhead), and prefers single-GPU placement (`split_mode=NONE`) with proportional multi-GPU splitting only as a fallback.
> 
> Integrates this allocation into `GGUFLanguageModel` and `GGUFEncoderModel` so model loads pass `main_gpu`/`split_mode`/`tensor_split`, and `GGUFLanguageModel` can re-compute `n_ctx` based on the selected GPU (or combined split VRAM) and re-run allocation when the context changes. Also updates `llamafarm_llama.Llama` to expose `split_mode`, improves memory reporting in `device.get_device_info()`, and adjusts context sizing to use `torch.cuda.mem_get_info()` (with optional `gpu_index`/override), with new unit tests covering allocator behavior and CUDA memory queries.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 8a5297afe0e935a7b2a28afef3baea49796cd767. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: runtimes/universal/models/gguf_language_model.py [hunk 1]
```diff
@@ -278,6 +280,54 @@ async def load(self) -> None:
...
+                            f"Context size adjusted for GPU {gpu_idx}: "
+                            f"{self.actual_n_ctx} -> {new_n_ctx}"
+                        )
+                        self.actual_n_ctx = new_n_ctx
+                        for w in new_warnings:
+                            logger.warning(w)
+            else:
+                logger.debug("No CUDA GPUs detected, using default GPU allocation")
+        except InsufficientVRAMError as e:
+            raise RuntimeError(str(e)) from e
```
[the remark this answers is not in the document]
Comment (BobbyRadford) [reply runtimes/universal/models/gguf_language_model.py:389]: this is not important

**Comment by mhamann** [comment]
This PR fixes #762.

**Review feedback addressed in latest commit (9a6dff5c):**
- Removed unused imports (`MagicMock`, `GPUAllocation`, `importlib`) from tests
- User-facing `RuntimeError` now uses a generic message; GPU hardware details are logged internally only
- Broke long `GPUDevice` constructor lines to comply with line length rules
- GPUs with <512 MiB free are now excluded from multi-GPU splits to prevent OOM on nearly-full small devices
- Encoder model passes explicit `n_ctx=512` to `Llama()` to match the VRAM estimate
- Context size recomputation only happens for single-GPU placement (`split_mode=NONE`), not multi-GPU splits
- Added test for tiny GPU exclusion from multi-GPU splits

**Review by BobbyRadford — APPROVED** [review]
